libs/ledger-live-common/src/device/use-cases/appDataBackup/backupAppData.ts

```diff
@@ -0,0 +1,69 @@
+import { AppStorageInfo, backupAppStorage, getAppStorageInfo } from "@ledgerhq/device-core";
+import Transport from "@ledgerhq/hw-transport";
+import { Observable, from, switchMap } from "rxjs";
+import { AppName, BackupAppDataError, BackupAppDataEvent, BackupAppDataEventType } from "./types";
+
+/**
+ * Backs up the application data for a specific app on a Ledger device.
+ * @param transport - The transport object used to communicate with the Ledger device.
+ * @param appName - The name of the application to backup.
+ * @returns A promise that resolves to the app storage data.
+ * @throws {BackupAppDataError}
+ */
+export function backupAppData(
+  transport: Transport,
+  appName: AppName,
+): Observable<BackupAppDataEvent> {
+  const obs = new Observable(subscriber => {
+    const sub = from(getAppStorageInfo(transport, appName))
+      .pipe(
+        switchMap(async (appDataInfo: AppStorageInfo) => {
+          subscriber.next({ type: BackupAppDataEventType.AppDataInfoFetched, data: appDataInfo });
+          const dataSize = appDataInfo.size;
+
+          // Check if there is any data to backup
+          if (dataSize === 0) {
+            subscriber.next({ type: BackupAppDataEventType.NoAppDataToBackup });
+            subscriber.complete();
+            return;
+          }
+
+          // Backup app data by chunks
+          let appData = Buffer.from([]);
+          let offset = 0;
+          while (offset < dataSize) {
+            const chunk: Buffer = await backupAppStorage(transport);
+            // Make sure that the process advances
+            if (chunk.length === 0) {
+              subscriber.error(new BackupAppDataError("Chunk data is empty."));
+              return;
+            }
+            appData = Buffer.concat([appData, chunk]);
+            offset += chunk.length;
+            subscriber.next({
+              type: BackupAppDataEventType.Progress,
+              data: Number((offset / dataSize).toFixed(2)),
+            });
+          }
+
+          // Check data size
+          if (appData.length !== dataSize) {
+            subscriber.error(new BackupAppDataError("App data size mismatch."));
+            return;
+          }
+
+          subscriber.next({
+            type: BackupAppDataEventType.AppDataBackedUp,
+            data: appData.toString("base64"),
+          });
+          subscriber.complete();
+        }),
+      )
+      .subscribe(subscriber);
+
+    return () => {
+      sub.unsubscribe();
+    };
+  });
+  return obs as Observable<BackupAppDataEvent>;
```

Comment:
[SHOULD] not use `as`, you can just do `const obs = new Observable<BackupAppDataEvent>(...)`

Reply:
It does not work, after some tries, this is the single solution in this case.

Reply:
I'm pretty sure there is a way to avoid using `as`, there must be a mistake somewhere if `new Observable<BackupAppDataEvent>(...)`.

I think the `subscribe(subscriber)` line 61 is what is incorrect because that observable you're subscribing to just returns `void` so your Observable is in fact an `Observable<BackupAppDataEvent | void>`.

You can just replace line 62 by `.subscribe()` and it works:

![Screenshot 2024-07-18 at 15 23 18](https://github.com/user-attachments/assets/45b04b68-ab1c-4243-b361-ed9bb72e5f5e)

Reply:
By the way by forcing with `as` and not removing the subscribe(subscriber) there was a bug: you were emitting `undefined` in your observable, which might cause a crash in the client as they're assuming that the type is correct.

Reply:
That's why `as` is really a trigger warning and we should always be very careful when using it. Sometimes there is no choice (but it's rare), and often when we use it, it just hides an underlying bug or at least some bad typing.

Reply:
Yes, I've found the solution. 👍